src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs

```diff
@@ -45,63 +45,77 @@ internal sealed class RegexParser
+        private ValueListBuilder<RegexOptions> _optionsStack;
 
-        private bool _ignoreNextParen = false;
+        private bool _ignoreNextParen; // flag to skip capturing a parentheses group
 
-        /*
-         * This static call constructs a RegexTree from a regular expression
-         * pattern string and an option string.
-         *
-         * The method creates, drives, and drops a parser instance.
-         */
-        public static RegexTree Parse(string re, RegexOptions op)
+        private RegexParser(string pattern, RegexOptions option, CultureInfo culture, Hashtable caps, int capsize, Hashtable capnames, Span<RegexOptions> optionSpan)
         {
-            RegexParser p;
-            RegexNode root;
-            string[] capnamelist;
-
-            p = new RegexParser((op & RegexOptions.CultureInvariant) != 0 ? CultureInfo.InvariantCulture : CultureInfo.CurrentCulture);
-
-            p._options = op;
+            Debug.Assert(pattern != null, "Pattern must be set");
+            Debug.Assert(culture != null, "Culture must be set");
 
-            p.SetPattern(re);
-            p.CountCaptures();
-            p.Reset(op);
-            root = p.ScanRegex();
+            _pattern = pattern;
+            _option = option;
+            _culture = culture;
+            _caps = caps;
+            _capsize = capsize;
+            _capnames = capnames;
 
-            if (p._capnamelist == null)
-                capnamelist = null;
-            else
-                capnamelist = p._capnamelist.ToArray();
+            _optionsStack = new ValueListBuilder<RegexOptions>(optionSpan);
+            _stack = default;
+            _group = default;
+            _alternation = default;
+            _concatenation = default;
+            _unit = default;
+            _currentPos = 0;
+            _autocap = default;
+            _capcount = default;
+            _captop = default;
+            _capnumlist = default;
+            _capnamelist = default;
+            _ignoreNextParen = false;
+        }
 
-            return new RegexTree(root, p._caps, p._capnumlist, p._captop, p._capnames, capnamelist, op);
+        private RegexParser(string pattern, RegexOptions option, CultureInfo culture, Span<RegexOptions> optionSpan)
+            : this(pattern, option, culture, new Hashtable(), default, null, optionSpan)
+        {
         }
 
-        /*
-         * This static call constructs a flat concatenation node given
-         * a replacement pattern.
-         */
-        public static RegexReplacement ParseReplacement(string rep, Hashtable caps, int capsize, Hashtable capnames, RegexOptions op)
+        public static RegexTree Parse(string pattern, RegexOptions option, CultureInfo culture)
         {
-            RegexParser p;
-            RegexNode root;
+            Span<RegexOptions> optionSpan = stackalloc RegexOptions[OptionStackDefaultSize];
+            var parser = new RegexParser(pattern, option, culture, optionSpan);
```

Comment:
Is the Regex parsing recursive?

If it is, is there a problem with consuming a lot more stack than before and running into stackoverflow?

Reply:
> Is the Regex parsing recursive?

The whole scanning logic happens in ScanRegex which then calls helper functions for different branches. There should not be any recursion involved in the parsing. Also we have a very good test coverage in S.T.RegularExpressions (> 90% code and branch coverage) and with https://github.com/dotnet/corefx/pull/29178 we added a lot of tests cases for the RegexParser.